kvserver/rangefeed: remove per-registration goroutines

wenyihu6 opened the pull request:
The goal of this PR is to remove the  _per registration_ goroutines that move
data from _per registration_ buffers to the client and replace it with a _per mux rangefeed_ goroutine 
managing a _per mux rangefeed_ buffer.

Since registrations are O(ranges) and since for large scale users, rangefeeds typically cover many
ranges on a node, this should reduce the number of goroutines needed to run a rangefeed in caught up mode.

In catch-up mode, the per-registration goroutine and buffered is still used.

This per-mux-rangefeed goroutine and buffer is managed by the new BufferedSender component.

Epic: None

Known follow-ups:

- [ ]  Perf test
- [ ]  Add buffered capacity as a follow up - add more tests related to memory accounting draining
- [ ]  Add test to cdc bench as a follow up
- [ ]  Enable the setting metamorphically as a follow up
- [ ]  Make catch up scan switch over test better

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125872)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver/rangefeed: add unbuffered registration

This patch adds unbufferedRegistration.

UnbufferedRegistration is like BufferedRegistration but uses BufferedSender to
buffer live raft updates instead of a using fixed size channel and having a
dedicated per-registration goroutine to volley events to underlying gRPC
stream. Instead, there is only one BufferedSender for each incoming
node.MuxRangefeed gRPC call. BufferedSender is responsible for buffering and
sending its updates to the underlying gRPC stream in a dedicated goroutine
O(node).

Resolved: #110432

Release note: A new cluster setting
`kv.rangefeed.buffered_stream_sender.enabled` can now be used to allow rangefeed
to use buffered sender for all rangefeed feeds instead of buffering events
separately per client per range.

Co-authored-by: Steven Danna danna@cockroachlabs.com

wenyihu6 committed:
kvserver/rangefeed: add node level buffered sender

This patch adds a node level BufferedSender which uses a queue
to buffer events before forwarding events to underlying grpc
stream.

Part of: #110432
Release note: note

Co-authored-by: Steven Danna danna@cockroachlabs.com

stevendanna committed:
kv/kvserver: remove redundant cancel func

Previously, UnbufferedSender maintained a list of cancel functions
that were canceled when processing an error. The goal of this
cancellation was to tear down running resource in the registration.

On error, something like the following would happen:

UnbufferedSender.SendBufferedError(err) would call cancel.  This
cancellation would cancel the context inherited by the registration's
runOutputLoop, causing it to exit. The registration would then call
disconnect, cancelling additional contexts, that also seem to be
intended to control the output loop, which is no longer
running. Disconnect would then send an error back to UnbufferedSender,
which would ignore it to prevent a duplicate sending of errors.

There are other error paths, where some of the components describe
above, aren't unnecessary, but, hopefully you can see that this is
ripe for some refactoring.

Here, we take a small step that we believe will also pave the way for
a clearer implementation of the BufferredSender.

Now, all error paths, start by calling
registration.Disconnect(err). The registration, cancels the context
controlling the runOutputLoop and sends an error to the stream, which
the stream then sends to the client.

There are some paths were Disconnect may be called twice still.

This simplification is made possible by one bit of indirection: The
processor now adds the registration to a map that lives on the stream
manager. This map will eventually also store any deferred cleanups
that the process requires when that stream is shutting down.

Note, we considered passing the registrations cleanup function back up
the call stack from register and may still explore that in the
future. However, it is not without challenge. Namely, the second we've
added the registration to the processor, it is possible that an error
is delivered from the processor that the stream manager would not yet
know how to handle.

We expect further naming and organizational improvements over the
coming commits.

Epic: none
Release note: None

Co-authored-by: Wenyi Hu wenyihu6@gmail.com

stevendanna committed:
kvserver/rangefeed: process unregister requests asyncronously

After the output loop completes, 2 cleanup steps are performed:

1) The registration is removed from the processors registry
2) The process is potentially stopped and removed from the replica.

Before this change, both of these happened syncronously after the
output loop finished.

With this change, step (1) happens asyncronously. To facilitate this,
an overflow mechanism is provided. This overflow mechanism potentially
allocates. Note that we expect that the number of requests is
relatively small and should be O(rangefeeds_on_range) so hopefully
this mechanism won't be used often.

Step (2) is now handled by the processor itself. After processing an
unregister request, if the set of registrations falls to zero, we
enqueue a Stop event for ourselves. Then, when processing the Stop, we
unregister ourselves from the replica.

Note that this may look like a small semantics change since the
previous unregister callback called Stop() which processes all events.
However, note that a Stopped event is processed after all other
events, so any events in the queue at the point of processing the
unregistration that enqueued the stop will be processed.

The motivation for this change is to eventually allow cleanup step (1)
to be run as part of registration.disconnect(), which needs to be
non-blocking. This is desired for a future change in which there is
not a dedicated goroutine to perform this cleanup.

Part of: #110432
Release note: note

stevendanna commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
```diff
@@ -0,0 +1,383 @@
+		ubr.metrics.RangeFeedCatchUpScanNanos.Inc(timeutil.Since(start).Nanoseconds())
+	}()
+
+	// In the future, we might want a separate queue for catch up scans.
```
Interesting, is there any reason we couldn't send the catchup scan to the buffer other than wanting to avoid overflowing it with catchup events. Might be worth filing an issue for.

wenyihu6 replied:
No, sending catch up scan updates to the buffer is fine correctness-wise. We made a choice here to make catch up scan bypass buffer here https://cockroachlabs.slack.com/archives/C05JB0QURND/p1724282080131549?thread_ts=1724072786.336579&cid=C05JB0QURND. It is also the only reason why we need to have send unbuffered on buffered stream.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/buffered_sender.go:
```diff
@@ -10,95 +10,117 @@ import (
+	bs.queueMu.Lock()
+	defer bs.queueMu.Unlock()
+	bs.queueMu.stopped = true
+	bs.queueMu.buffer.free()
```
Nice catch, done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/processor_helpers_test.go:
```diff
@@ -233,18 +259,42 @@ func (h *processorTestHelper) triggerTxnPushUntilPushed(t *testing.T, pushedC <-
+		return "scheduled processor with buffered registration"
+	case scheduledProcessorWithUnbufferedReg:
+		return "scheduled processor with unbuffered registration"
+	}
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/processor_helpers_test.go:
```diff
@@ -233,18 +259,42 @@ func (h *processorTestHelper) triggerTxnPushUntilPushed(t *testing.T, pushedC <-
 
-var testTypes = []rangefeedTestType{}
+type rangefeedTestType struct {
+	processorType procType
```
Changed it to useBufferedSender bool - https://github.com/cockroachdb/cockroach/pull/133345/files.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream_manager.go:
```diff
@@ -0,0 +1,209 @@
+	// directly. Note that this call may block.
+	sendUnbuffered(ev *kvpb.MuxRangeFeedEvent) error
+	// sendBuffered buffers a RangeFeedEvent before sending to the underlying grpc
+	// stream. It should be unblocking.
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream_manager.go:
```diff
@@ -0,0 +1,209 @@
+func (sm *StreamManager) DisconnectStream(streamID int64, err *kvpb.Error) {
+	if err == nil {
+		log.Fatalf(context.Background(),
+			"unexpected: DisconnectStream called with non-error event")
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream_manager.go:
```diff
@@ -0,0 +1,209 @@
+	defer sm.streams.Unlock()
+	if d.IsDisconnected() {
+		// If the stream is already disconnected, we don't add it to the stream. An
+		// error will be sent back to the client.
```
Done.

stevendanna replied:
It depends on what we mean by "client" here. A more accurate statement is probably "The registration will have already send an error to the sender" perhaps?

wenyihu6 commented on pkg/kv/kvserver/rangefeed/buffered_sender.go:
```diff
@@ -10,95 +10,117 @@ import (
-	// Disconnect stream and cancel context. Then call SendBuffered with the error
-	// event.
-	panic("unimplemented: buffered sender for rangefeed #126560")
+// sendBuffered buffers the event before sending it to the underlying grpc
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/buffered_sender.go:
```diff
@@ -10,95 +10,117 @@ import (
+// sendBuffered buffers the event before sending it to the underlying grpc
+// stream. It should not block since errors are sent directly here. sendBuffered
+// will take the ownership of the alloc and release it if the return error is
+// non-nil.
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/buffered_sender.go:
```diff
@@ -10,95 +10,117 @@ import (
+	if bs.queueMu.stopped {
+		return errors.New("stream sender is stopped")
+	}
+	alloc.Use(context.Background())
```
Sg, I will file an issue about this.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/buffered_sender.go:
```diff
@@ -10,95 +10,117 @@ import (
-		sender:  sender,
-		metrics: metrics,
+	// queueMu protects the buffer queue.
+	queueMu struct {
+		syncutil.Mutex
+		stopped bool
+		buffer  *eventQueue
 	}
```
I will add it in a follow up PR. I didn't it to this PR since I wasn't sure what cap size would be good. I don't want this to be the reason we are not seeing good performance. And we already have memory accounting for these events on the server side.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
```diff
@@ -0,0 +1,383 @@
+// b) Processor level clean-up: p.reg.Unregister (ubr.cleanup)
+//
+// c) BufferedSender level clean-up: stream.SendError
+//   - metrics updates, send a disconnect error back to rangefeed client
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
```diff
@@ -0,0 +1,383 @@
+	ubr.mu.Unlock()
+
+	if err := ubr.maybeRunCatchUpScan(ctx); err != nil {
+		// Important to disconnect before draining.
```
Renamed to drainCatchUpBuffer.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
```diff
@@ -0,0 +1,383 @@
+	defer ubr.mu.Unlock()
+	// TODO(wenyihu6): Check if we can just discard without holding the lock first
+	// ? We shouldn't be reading from the buffer at the same time
+	ubr.discardCatchUpBufferWithLock(ctx)
```
Changed it to inline. Good idea about the assertion added.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/processor_test.go:
```diff
@@ -300,24 +300,28 @@ func TestProcessorBasic(t *testing.T) {
 		// Cancel the first registration.
-		r1Stream.Cancel()
+		if rt == scheduledProcessorWithUnbufferedSender {
+			r1Stream.Cancel()
```
Test Cancel for buffered registration and disconnect for unbuf

wenyihu6 commented on pkg/kv/kvserver/rangefeed/processor_test.go:
```diff
@@ -424,106 +428,110 @@ func TestProcessorOmitRemote(t *testing.T) {
-		defer func() {
-			if unblock != nil {
-				unblock()
+	testutils.RunValues(t, "feed type", []rangefeedTestType{scheduledProcessorWithUnbufferedSender},
```
Add a test for buffered sender capacity

wenyihu6 commented on pkg/kv/kvserver/rangefeed/processor_helpers_test.go:
```diff
@@ -23,6 +23,31 @@ import (
+		// In practise, StreamMuxer is responsible for invoking cleanup in the
+		// background. For simplicity, we call it directly here.
+		go tb.cleanup()
+		time.Sleep(10 * time.Millisecond)
```
clean up time.Sleep

wenyihu6 commented:
Pulled the first commit into https://github.com/cockroachdb/cockroach/pull/133685.

tbg reviewed (changes requested):
I gave this a first pass. Even though I understand the high-level picture, there is too much here, especially in the first commit, to really review in depth; entire files have moved and Reviewable is already extremely confused and this would only get worse. I would like to suggest breaking the first commit up into a sequence of PRs that we can then iterate and improve on, which should now be possible since we know how it should all shake out.
While doing so, I would try to look at ways to avoid the "obfuscation" that arises from sharing the `sender` interface, and avoid code movement except in commits that don't make any semantic changes. My (perhaps naive) suggestion would be:

1. extract the `StreamManager`. (There is no buffered stream yet, so it's used only for the unbuffered stream). (Extract it in place, then move it to new file). We merge that.
2. Do the disconnector refactor. (If this can be split up further, great).
3. now unless I've missed something, all the cleanup has happened and we can introduce the unbuffered sender (i.e. we pick up the sequence of commits you have here). First commit we devise a way to avoid sharing the `sender` interface. Then introduce the interface for the buffered sender. Make sure StreamManager can route it. (buffered sender still unimplemented here, just have a stub). Then replace the stub with the real thing, add tests, etc.
4. add unbuffered registration.

WDYT?

> Reviewed 7 of 24 files at r14, 20 of 20 files at r18, 7 of 7 files at r19, 8 of 8 files at r20, 12 of 12 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-:-OBPs4voE6yQKQwdhh5g:b-afsiw)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @stevendanna and @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 72 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPpVyo-Kj0ZaxqEuvv:-OBPpVyo-Kj0ZaxqEuvw:b-iux1yk) ([raw file](https://github.com/cockroachdb/cockroach/blob/093e23a09e7d204382a5687ed85676ceb068102f/pkg/kv/kvserver/rangefeed/buffered_sender.go#L72)):*
> ```Go
> 
> // sendUnbuffered sends the event directly to the underlying ServerStreamSender.
> // It bypasses the buffer.
> ```

But this doesn't seem to be called (`PerRangeEventSink.SendUnbuffered` calls it, but this sink wouldn't wrap a buffered sender I think).

So we could drop this method from here and the `sender` interface and instead make an interface-widening assertion only in the case of the unbuffered sender.
___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 96 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPptnr0xyxwO045Qwv:-OBPptnr0xyxwO045Qww:b-sdccu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/093e23a09e7d204382a5687ed85676ceb068102f/pkg/kv/kvserver/rangefeed/buffered_sender.go#L96)):*
> ```Go
> 			for {
> 				e, success := bs.popFront()
> 				if success {
> ```

nit:

```go
if !success {
  break
}
// do the happy path
```
___
*[`pkg/kv/kvserver/rangefeed/stream_manager.go` line 21 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPgTwm5y9fUAiFgT2U:-OBPgTwnDI1ejwNYJk7f:b6shq88) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebd85b709605c6f52f5f34feddcbb013dc1ae35/pkg/kv/kvserver/rangefeed/stream_manager.go#L21)):*
> ```Go
> // +-----------------+               +-----------------+                  +-----------------+
> // |                 | Disconnect    |                 |  r.disconnect    |                 |
> // |  MuxRangeFeed   +-------------->|   SteamManager  +----------------->|   Registration  |
> ```

StreamManager
___
*[`pkg/kv/kvserver/rangefeed/stream_manager.go` line 49 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPgnLn2srEopeTC53V:-OBPgnLoDwKh1NV1mEDz:bb7sbnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebd85b709605c6f52f5f34feddcbb013dc1ae35/pkg/kv/kvserver/rangefeed/stream_manager.go#L49)):*
> ```Go
> 	errCh chan error
> 
> 	// Note that lockedMuxStream wraps the underlying grpc server stream, ensuring
> ```

Patch up the comment to explain that sender is thread-safe (which I think is what it is now saying). I think it's helpful to also mention the possible implementation (as the comment does now).
___
*[`pkg/kv/kvserver/rangefeed/stream_manager.go` line 108 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPhJgK60IsNNQ1MGIz:-OBPhJgK60IsNNQ1MGJ-:b-2nyg8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebd85b709605c6f52f5f34feddcbb013dc1ae35/pkg/kv/kvserver/rangefeed/stream_manager.go#L108)):*
> ```Go
> // OnError is a callback that is called when a sender sends a rangefeed
> // completion error back to the client. Note that we check for the existence of
> // streamID to avoid metrics inaccuracy when the error is sent before the stream
> ```

What happens in this case when the stream is added? Why isn't a stream always added before it can generate errors?
___
*[`pkg/kv/kvserver/rangefeed/stream_manager.go` line 123 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPjq_f5jvfp8qSUdxA:-OBPjq_f5jvfp8qSUdxB:b-4k2fue) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebd85b709605c6f52f5f34feddcbb013dc1ae35/pkg/kv/kvserver/rangefeed/stream_manager.go#L123)):*
> ```Go
> 	if err == nil {
> 		log.Fatalf(context.Background(),
> 			"unexpected: DisconnectStream called with non-nil error")
> ```

nil, not non-nil.
___
*[`pkg/kv/kvserver/rangefeed/unbuffered_sender.go` line 129 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPl61B3-p89PNxFETB:-OBPl61B3-p89PNxFETC:b8jmlws) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebd85b709605c6f52f5f34feddcbb013dc1ae35/pkg/kv/kvserver/rangefeed/unbuffered_sender.go#L129)):*
> ```Go
> // needs to make sure this is called only with non-nil error events. Important
> // to be thread-safe.
> func (ubs *UnbufferedSender) sendBuffered(
> ```

it's surprising to me that this method is used only in one specific case and the sibling on `BufferedSender` is the main workhorse. I think this is because one interface is used for two different systems: an `UnbufferedSender` is used by a `PerRangeEventSink` and a `BufferedSender` is used by a `BufferedPerRangeEventSink`.

I find that keeping legacy impls around tends to greatly obfuscate the code. I am looking at this code:

```
	case *BufferedSender:
		return &BufferedPerRangeEventSink{
			PerRangeEventSink: NewPerRangeEventSink(rangeID, streamID, sender),
		}
	case *UnbufferedSender:
		return NewPerRangeEventSink(rangeID, streamID, sender)
	default:
```

and I wonder why we need to use the same `sender` interface in both. Can't they take different interfaces? Or can we at least make the interface "canonical" for the non-legacy (buffered) case and then just type assert in the legacy case and use the underlying type (a non-buffering sender in the unbuffered case)?

This suggestion is not for this PR. But anything that helps make this code more accessible will help for the future.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125872)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/stream_manager.go` line 108 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPhJgK60IsNNQ1MGIz:-OBQ2AgyBn2h-CJuLW8G:b-81ar1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebd85b709605c6f52f5f34feddcbb013dc1ae35/pkg/kv/kvserver/rangefeed/stream_manager.go#L108)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

What happens in this case when the stream is added? Why isn't a stream always added before it can generate errors?
</blockquote></details>

Just as a heads up: This has been one of the main points of back and forth in some past review.

A registration can send us error once it is registered with a processor because of an error at the processor level or from the catch-up scan.  This can happen before the Disconnector has been sent back up the stack to us.  

Previous iterations solved this by having the registration process register the callback with us by sending something down the stack and then setting the callback at the point of registration construction. No one was particularly happy with the complexity this appeared to add.

In this iteration, we've opted to send this Disconnector back up the stack and then solve the race by checking whether the stream is Disconnected before adding it to the map. Because we only increment stats when we added the stream to the map, we only want to decrement them on streams actually deleted from the map.


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TYFTR! I started working with Steven to break down the PRs into smaller chunks. I've already started a series here (https://github.com/cockroachdb/cockroach/pull/134953, https://github.com/cockroachdb/cockroach/pull/134954, https://github.com/cockroachdb/cockroach/pull/134955, https://github.com/cockroachdb/cockroach/pull/134956, https://github.com/cockroachdb/cockroach/pull/134957, https://github.com/cockroachdb/cockroach/pull/134958, https://github.com/cockroachdb/cockroach/pull/134959). The last one is still a bit large, but I'll see if I can split it further - it will require a bit more surgery. Just wanted to get the ball rolling on the first few ones. Let me know if this is along the lines you were thinking. 





<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125872)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @stevendanna and @tbg)

___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 72 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPpVyo-Kj0ZaxqEuvv:-OB_wvKl9ddIOBc-KIev:b-sz2o9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/093e23a09e7d204382a5687ed85676ceb068102f/pkg/kv/kvserver/rangefeed/buffered_sender.go#L72)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

But this doesn't seem to be called (`PerRangeEventSink.SendUnbuffered` calls it, but this sink wouldn't wrap a buffered sender I think).

So we could drop this method from here and the `sender` interface and instead make an interface-widening assertion only in the case of the unbuffered sender.
</blockquote></details>

We can discuss the contracts more in https://github.com/cockroachdb/cockroach/pull/134957 and see if there is a cleaner way to do things. It is a bit confusing since we allow `PerRangeEventSink` to wrap a `BufferedSender` https://github.com/cockroachdb/cockroach/blob/6478174dde4e64aa6e5c2a2d0e61e775e897626c/pkg/kv/kvserver/rangefeed/stream_manager.go#L99.  More details in https://github.com/cockroachdb/cockroach/pull/134957#issuecomment-2471737003. iirc, the only reason we need both `PerRangeEventSink` and `BufferedPerRangeEventSink` is because we want to have both `BufferedStream` and `Stream` interfaces to tell `p.Register` which registration type to initialize. 
___
*[`pkg/kv/kvserver/rangefeed/buffered_sender.go` line 96 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPptnr0xyxwO045Qwv:-OB_yWD28Ro57yKcLA3K:b-4igljb) ([raw file](https://github.com/cockroachdb/cockroach/blob/093e23a09e7d204382a5687ed85676ceb068102f/pkg/kv/kvserver/rangefeed/buffered_sender.go#L96)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit:

```go
if !success {
  break
}
// do the happy path
```
</blockquote></details>

I will try to remember this when we are adding the BufferedSender PR.
___
*[`pkg/kv/kvserver/rangefeed/stream_manager.go` line 21 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPgTwm5y9fUAiFgT2U:-OB_y7POAVrfcanY88qe:b-snynw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebd85b709605c6f52f5f34feddcbb013dc1ae35/pkg/kv/kvserver/rangefeed/stream_manager.go#L21)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

StreamManager
</blockquote></details>

Done in https://github.com/cockroachdb/cockroach/blob/39afe4eeae0ef4161930d6525fec42b18f85ba56/pkg/kv/kvserver/rangefeed/stream_manager.go#L21.
___
*[`pkg/kv/kvserver/rangefeed/stream_manager.go` line 49 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPgnLn2srEopeTC53V:-OB_yDw87EpNtD-VeyBp:bvpktyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebd85b709605c6f52f5f34feddcbb013dc1ae35/pkg/kv/kvserver/rangefeed/stream_manager.go#L49)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Patch up the comment to explain that sender is thread-safe (which I think is what it is now saying). I think it's helpful to also mention the possible implementation (as the comment does now).
</blockquote></details>

Done in https://github.com/cockroachdb/cockroach/blob/39afe4eeae0ef4161930d6525fec42b18f85ba56/pkg/kv/kvserver/rangefeed/stream_manager.go#L49-L50. I added a contract method to the sender interface to make this clearer. 
___
*[`pkg/kv/kvserver/rangefeed/stream_manager.go` line 123 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPjq_f5jvfp8qSUdxA:-OB_yLBP4FO5UmdA1xdb:b84s0b3) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebd85b709605c6f52f5f34feddcbb013dc1ae35/pkg/kv/kvserver/rangefeed/stream_manager.go#L123)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nil, not non-nil.
</blockquote></details>

Oops, done in https://github.com/cockroachdb/cockroach/blob/39afe4eeae0ef4161930d6525fec42b18f85ba56/pkg/kv/kvserver/rangefeed/stream_manager.go#L117. 
___
*[`pkg/kv/kvserver/rangefeed/unbuffered_sender.go` line 129 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/125872#-OBPl61B3-p89PNxFETB:-OB_yQZ4Ei4ZPhq8BbAV:b8xomze) ([raw file](https://github.com/cockroachdb/cockroach/blob/aebd85b709605c6f52f5f34feddcbb013dc1ae35/pkg/kv/kvserver/rangefeed/unbuffered_sender.go#L129)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

it's surprising to me that this method is used only in one specific case and the sibling on `BufferedSender` is the main workhorse. I think this is because one interface is used for two different systems: an `UnbufferedSender` is used by a `PerRangeEventSink` and a `BufferedSender` is used by a `BufferedPerRangeEventSink`.

I find that keeping legacy impls around tends to greatly obfuscate the code. I am looking at this code:

```
	case *BufferedSender:
		return &BufferedPerRangeEventSink{
			PerRangeEventSink: NewPerRangeEventSink(rangeID, streamID, sender),
		}
	case *UnbufferedSender:
		return NewPerRangeEventSink(rangeID, streamID, sender)
	default:
```

and I wonder why we need to use the same `sender` interface in both. Can't they take different interfaces? Or can we at least make the interface "canonical" for the non-legacy (buffered) case and then just type assert in the legacy case and use the underlying type (a non-buffering sender in the unbuffered case)?

This suggestion is not for this PR. But anything that helps make this code more accessible will help for the future.
</blockquote></details>

We can discuss this more in https://github.com/cockroachdb/cockroach/pull/134957. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
Closing in favour of https://github.com/cockroachdb/cockroach/pull/136310.
